src/main/java/org/opensearch/neuralsearch/processor/optimization/InferenceFilter.java

```diff
@@ -0,0 +1,233 @@
+                filteredProcessMap.put(key, filteredInnerMap.isEmpty() ? null : filteredInnerMap);
+            } else if (value instanceof List) {
+                List<Object> processedList = filterListValue(
+                    currentPath,
+                    (List<Object>) value,
+                    sourceAndMetadataMap,
+                    existingSourceAndMetadataMap
+                );
+                filteredProcessMap.put(key, processedList);
+            } else {
+                Object processedValue = filterInferenceValue(currentPath, value, sourceAndMetadataMap, existingSourceAndMetadataMap, -1);
+                filteredProcessMap.put(key, processedValue);
+            }
+        }
+        return filteredProcessMap;
+    }
+
+    /**
+     * Processes a list of values by comparing them against source and existing metadata.
+     *
+     * @param embeddingKey The current path in dot notation for the list being processed
+     * @param processList The list of values to process
+     * @param sourceAndMetadataMap SourceAndMetadataMap of ingestDocument Document
+     * @param existingSourceAndMetadataMap SourceAndMetadataMap of existing Document
+     * @return A processed list containing non-filtered elements
+     */
+    protected List<Object> filterListValue(
+        String embeddingKey,
+        List<Object> processList,
+        Map<String, Object> sourceAndMetadataMap,
+        Map<String, Object> existingSourceAndMetadataMap
+    ) {
+        String textKey = reversedFieldMap.get(embeddingKey);
+        if (Objects.isNull(textKey)) {
+            return processList;
+        }
+        Optional<Object> existingList = ProcessorUtils.getValueFromSource(existingSourceAndMetadataMap, textKey);
+        if (existingList.isPresent() == false) {
+            // if list is not present with given text key, process each entry in processList individually
+            return filterMapValuesInList(processList, embeddingKey, sourceAndMetadataMap, existingSourceAndMetadataMap);
+        }
+        // if list is present with given textKey, but is not a list type, no valid comparison can be made, return processList
+        if (existingList.get() instanceof List == false) {
+            return processList;
+        }
+        // retrieve embedding for given embedding key, if not present or is not a list type, return processList, as no embedding can be
+        // copied over
+        Optional<Object> embeddingList = ProcessorUtils.getValueFromSource(existingSourceAndMetadataMap, embeddingKey);
+        if (embeddingList.isPresent() == false || embeddingList.get() instanceof List == false) {
+            return processList;
+        }
+        // return empty list if processList and existingList are equal and embeddings are copied, return empty list otherwise
+        return filterInferenceValuesInList(
+            processList,
+            (List<Object>) existingList.get(),
```

Comment:
Same for this. Not all List is List<Object>. For example, we cannot cast List<Integer> to List<Object>